File: src/ruby/longest_common_prefix.rb
def longest_common_prefix(strs)
  if strs != []
    prefix = strs.min_by(&:length)
    while prefix != ""
      return prefix if strs.all?{|str| str[0...prefix.length] == prefix}
      prefix = prefix.chop
    end
  end
  return ""
end

p longest_common_prefix(["a", "ab","abcd"]) # "a"
p longest_common_prefix(["", "abc", "abc"]) # ""
p longest_common_prefix(["ab", "ab"]) # "ab"
p longest_common_prefix(["abc", "cabc"]) # ""
p longest_common_prefix([]) # ""

File: src/python/src/longest_common_prefix.py
# https://leetcode.com/problems/longest-common-prefix/description/?envType=company&envId=facebook&favoriteSlug=facebook-thirty-days


class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        shortest_len = min([len(w) for w in strs])
        for i in range(shortest_len):
            ref = strs[0][i]
            if not all(w[i] == ref for w in strs):
                return strs[0][:i]
        return strs[0][:shortest_len]
